### PR TITLE
Fix return type of Header::all

### PR DESCRIPTION
**Pull request by Seldaek**
| Q             | A
| ------------- | ---
| Branch?       | 5.4
| Bug fix?      | yes
| New feature?  | no 
| Deprecations? | no 
| Tickets       | Fix https://github.com/phpstan/phpstan-symfony/issues/264
| License       | MIT
| Doc PR        | symfony/symfony-docs#... 

cc @derrabus

### CODE REVIEW

**Review by stof — APPROVED** [review]

File: src/Symfony/Component/HttpFoundation/HeaderBag.php [hunk 1]
```diff
@@ -65,7 +65,7 @@ public function __toString()
      *
      * @param string|null $key The name of the headers to return or null to get them all
      *
-     * @return array<string, array<int, string|null>>|array<int, string|null>
+     * @return $key is null ? array<string, array<int, string|null>> : array<int, string|null>
```
Comment (VincentLanglet) [inline src/Symfony/Component/HttpFoundation/HeaderBag.php]:
```suggestion
     * @return ($key is null ? array<string, array<int, string|null>> : array<int, string|null>)
```
See https://phpstan.org/r/18e18b32-1cc3-4647-875b-c756ad1ad4d7

Also, to keep IDE support I would recommend
```suggestion
     * @return array<string, array<int, string|null>>|array<int, string|null>
     * @phpstan-return ($key is null ? array<string, array<int, string|null>> : array<int, string|null>)
```
(or psalm-return if preferred)

**Review by VincentLanglet — CHANGES_REQUESTED** [review]

File: src/Symfony/Component/HttpFoundation/HeaderBag.php (hunk 1 above)
Comment (Seldaek) [reply src/Symfony/Component/HttpFoundation/HeaderBag.php]: Applied the first fix, thanks.

The second I would agree but I think it's again symfony policy so leaving it as is.

**Comment by derrabus** [comment]
AFAIK (cc @nicolas-grekas) we only annotate types that have broad tooling support. This is certainly not the case for this kind of conditional types. And we don't use vendor-prefixed annotations either.

If that issue could be solved by the Symfony plugin for Psalm, I think I'd prefer that way.

**Comment by VincentLanglet** [comment]
> If that issue could be solved by the Symfony plugin for Psalm, I think I'd prefer that way.

Stub with this conditionnal return type can be added in both phpstan-symfony and psalm-symfony plugin yes.

> AFAIK (cc @nicolas-grekas) we only annotate types that have broad tooling support. This is certainly not the case for this kind of conditional types. And we don't use vendor-prefixed annotations either.

Is there any plan to start using prefixed annotations later ? Or what is the reason to not to ?

**Comment by Seldaek** [comment]
@nicolas-grekas IMO this policy is a mistake at this point. It might have made sense a while back but PHPStan usage is spreading more and more, and having type annotations shipped with the framework would make more sense. It ensures the types are always in sync with the Symfony version and they would also benefit Symfony itself in terms of added static analysis capability. 

I've seen a similar thing happen in the TypeScript ecosystem where at first typings packages where provided to type things from the outside, and then eventually all major packages started shipping their own type definitions or even switching their source completely to typescript.

PHPStan is now a very established tool, and Psalm also reads `@phpstan-x` annotations (and vice-versa), and PHPStorm is adding support for more and more types in every release, so I think the "broad tooling support" argument is maybe moot?

Just my 2c :)

**Comment by derrabus** [comment]
> Is there any plan to start using prefixed annotations later ?

There are no plans yet.

> Or what is the reason to not to ?

We use Psalm merely as a code review tool these days. If you want to dicuss an initiaive to make the codebase more accessible for SA tools, it would maybe be better to discuss this in a separate issue first.

**Comment by Seldaek** [comment]
Yeah, I don't have time for that crusade, so closing here, moved to https://github.com/phpstan/phpstan-symfony/pull/282 - but for the record my point above stands.

**Comment by nicolas-grekas** [comment]
For the record, having phpstan+psalm+phpstorm support for the syntax is our current bar.

**Comment by VincentLanglet** [comment]
> For the record, having phpstan+psalm+phpstorm support for the syntax is our current bar.

Since PHPStorm 2022.3.1, the conditional return is supported by PHPStorm.
https://youtrack.jetbrains.com/issue/WI-66465/Support-parsing-of-PhpStans-conditional-return-types

So would the PR be accepted if re-opened ? @nicolas-grekas

**Comment by ro0NL** [comment]
do  phpstan+psalm share same syntax now? IMHO it's essential for maintenance. Unless it's not, and we choose vendor prefixes, for what is a single language :D

**Comment by nicolas-grekas** [comment]
Please resubmit after checking with the main tools yes.

**Comment by VincentLanglet** [comment]
Sure, https://github.com/symfony/symfony/pull/49007

**Pull request closed without merging**